System for Creating and Storing Custom Messages

leggettc18 opened the pull request:
Custom Message Manager capable of managing multiple message tables, so each Mod/Feature requiring custom text can have their own and avoid conflicts with the base game's and other mods' message tables. Will be used pretty heavily by rando and also in use to override some base game text when inject skulltula count is enabled. Also comes with a way to move the existing custom message logic out of `z_message_PAL.c` to keep it closer in line with decomp. Original message with more details below.

This work originally started out on #767. I had set out to create essentially some helper functions for rando to make custom messages for some of the custom items we'll need to make later (bean pack, dungeon-specific small keys, etc.). As I continued to build it out over the weekend, I realized it could have uses outside of rando, and it might be better to store the messages somewhere instead of creating them on the fly and immediately dropping them. This should also be able to be hooked into by a scripting engine for modders later on as well without too much work, I kept that in mind while developing.

The idea is that a class instace, either created by the future scripting engine or coded into the main codebase, will create a message table and store their messages as part of their initialization process. Currently I have this set up as a singleton class that manages all the message tables, so that a mod can just somehow provide an ID, perhaps through `globalContext`, to tell the game which message table to use. But I can also see the benefits to letting each class instance that needs one to store their own `CustomMessage` instance (or instances if desired).

TODO:
- [x] Make a `CreateMessage` function for generic text that isn't GetItem text.
- [x] Provide a way to specify the text box type and placement.
- [ ] More symbol-replacement options for things like auto-dismissal of text boxes and centering text? (May be better to just expect people to use the control codes).
- [x] Documentation.
- [x] Implementing any suggestions.

EDITED ON 7/19/2022 FOR BREVITY.

leggettc18 committed:
Early version of custom messages.

leggettc18 committed:
Makes sure custom get item textboxes match.

Specifically, they are all bottom of the screen and blue.

leggettc18 committed:
Refactors custom message code to its own file

leggettc18 committed:
Renames the separated custom messages files to prevent conflict.

leggettc18 committed:
Merge branch 'develop' of https://github.com/HarbourMasters/Shipwright into rando-custom-messages

leggettc18 committed:
Merge branch 'develop' of https://github.com/HarbourMasters/Shipwright into rando-custom-messages

leggettc18 committed:
Fixes newline function to return correct character.

leggettc18 committed:
Merge branch 'develop' of https://github.com/HarbourMasters/Shipwright into rando-custom-messages

leggettc18 committed:
Added an extra helper function and removed instant text control code from ITEM_OBTAINED

leggettc18 committed:
Merge branch 'develop' of https://github.com/HarbourMasters/Shipwright into rando-custom-messages

leggettc18 committed:
Adds text for Bottle with Blue Fire as a demo.

leggettc18 committed:
Refactors CustomMessages to not be rando specific.

Pulls the Custom Message related code out into it's own class, which has an initialization phase where other enhancements / future features can create messages during an initialization phase to be stored and retrieved later. Along with this refactoring, the 4 bottle messages from the previous rando-specific system are now created and stored during intialization and retrieved by their getItemId.

Now that it isn't rando specific, the goal is to move anything text changes that are hard-coded into z_message_PAL.c and refactor it so that future text additions/overrides can be done without modifying that file.

leggettc18 committed:
Does some cleanup of unecessary code and cases of no string provided.

leggettc18 committed:
Implements function to return custom message if one exists.

`CustomMessage_RetrieveIfExists` is placed at the very beginning of the z_message_PAL.c `Message_OpenText` if statement after any special changes to textIds are made. This function will either return either true or false and if true it will populate the necessary values of font and msgCtx to display said message. It's placement at the beginning also allows for overriding existing textIds to have new information, which will come in handy later.

leggettc18 committed:
Simplifies the function signature of `CustomMessage_RetrieveIfExists`.

leggettc18 committed:
more cleanup

leggettc18 committed:
Implements adding message tables and retrieving by an ID.

Basically, some external code can choos a unique id and create a message table for itself. The idea is that modders can hook into this as well so they can get their own message table and don't have to worry about conflicting with the base game's textIDs. I have also implemented this setup for Randomizer so that the pre-filled bottles (which didn't exist in the original outside of Ruto's letter) have unique text as compared to the text for the item they are filled with.

leggettc18 committed:
Implements the custom text for gInjectSkulltulaCount.

Also auto-dissmisses the textbox when gSkulltulaFreeze is on. When gInjectSkulltulaCount is off, text is not auto-dismissed just like before, since that is base game text still.

leggettc18 committed:
Implements a basic CreateMessage function for non-GetItem text.

leggettc18 commented:
I'm assuming this is not getting into Rachael Bravo so I'm changing the base branch on this one to develop-zhora.

leggettc18 committed:
Merge in develop-zhora

leggettc18 committed:
Adds ability to store and retrieve textbox size and position.

leggettc18 committed:
Cleanup and move most hint logic to CustomMessage.

Haven't removed where rando stores the hints for itself and the ganon and altar hints are still pulled from the randomizer class' local storage, but gossip stone hints are pulled from the custom message table now.

leggettc18 committed:
Removes randomizer_custom_messages, as I relocated its logic elsewhere.

leggettc18 committed:
Merge branch 'develop-zhora' of https://github.com/HarbourMasters/Shipwright into custom-messages

leggettc18 committed:
Does some refactoring of rando custom message registration

The goal was less verbosity when adding new get-item messages later.

leggettc18 committed:
Fixes minor oversight from previous commit.

leggettc18 committed:
Fixes storage of altar and ganon hints/text.

Only one language is stored in gSaveContext at the moment so store the one language in all 3 CustomMessageEntry languages.

leggettc18 committed:
Pulls altar text from CustomMessageTable instead of gSaveContext.

leggettc18 committed:
Cleans up unused z_message_PAL code.

leggettc18 committed:
Takes Ganondor's text & hint from CustomMessageTable

leggettc18 committed:
Moves deku scrub messages into CustomMessageTable.

leggettc18 committed:
Merge in develop-zhora

leggettc18 committed:
Puts CustomMessage files in filters in Visual Studio

leggettc18 committed:
Renames CustomMessage to CustomMessageManager

leggettc18 commented:
Alright, I think I finally got this to a point where I'm happy enough with it for it to be reviewed. This PR includes the custom message manager and the necessary changes to put Randomizer's existing custom text and some "vanilla" enhancement's custom text out of z_message_PAL.c and into custom message tables instead.

Critiques welcome!

briaguya0 commented on soh/include/z64.h:
```diff
@@ -30,6 +30,9 @@
 #include "ichain.h"
 #include "regs.h"
 
+#define NOGDI
+#define WIN32_LEAN_AND_MEAN
```
would you mind making a quick little pr with this only? could be nice to have a record of what it fixes

leggettc18 replied:
Oh I might be able to remove that actually. It was also only an issue with certain `#includes` in certain files, so there may be a better way to fix it. Let me look into that again.

leggettc18 replied:
Ah, I remember now, without those two lines there I had issues with _Polygon getting redefined.

```
error C2365: '_Polygon': redefinition; previous definition was 'typedef' (compiling source file soh\Enhancements\custom_message\CustomMessageManager.cpp)
1>C:\Users\legge\Documents\Projects\Shipwright\soh\include\z64.h(899): message : see declaration of '_Polygon' (compiling source file soh\Enhancements\custom_message\CustomMessageManager.cpp)
```

And an identical error for randomizer.cpp. Putting that in z64.h didn't seem to cause any problems from my testing and fixed both errors. Unless there's a better way to fix that that you're aware of?

leggettc18 replied:
I've found I can also fix the _Polygon redefinition by putting those two lines in message_data_static.h instead. Experimenting to see if there's any better fix as well.

leggettc18 replied:
I think I've fixed this in a better way now. Basically there was a complex web of `#includes` that caused a `_Polygon` type to be included from both `wingdi.h` (via `Windows.h` in OTRGlobals.cpp) and `z64.h` (seems like it's coming in via `message_data_static.h` through a few other chained `#includes`). I moved the highlighted code into `CustomMessageManager.h` before the `#include message_data_static.h` and that allows the code to compile, and puts this new change into a file relevant to this PR.

briaguya0 commented on soh/soh/Enhancements/custom_message/CustomMessageManager.cpp:
```diff
@@ -0,0 +1,150 @@
+    }
+}
+
+void CustomMessageManager::FormatCustomMessage(std::string& message, ItemID iid) {
+    message.insert(0, ITEM_OBTAINED(iid));
+    size_t start_pos = 0;
+    std::replace(message.begin(), message.end(), '&', NEWLINE()[0]);
+    while ((start_pos = message.find('^', start_pos)) != std::string::npos) {
+        message.replace(start_pos, 1, WAIT_FOR_INPUT() + ITEM_OBTAINED(iid));
+        start_pos += 3;
+    }
+    std::replace(message.begin(), message.end(), '@', PLAYER_NAME()[0]);
+    ReplaceSpecialCharacters(message);
+    ReplaceColors(message);
+    message += MESSAGE_END();
+}
+
+void CustomMessageManager::FormatCustomMessage(std::string& message) {
+    size_t start_pos = 0;
+    std::replace(message.begin(), message.end(), '&', NEWLINE()[0]);
+    std::replace(message.begin(), message.end(), '^', WAIT_FOR_INPUT()[0]);
+    std::replace(message.begin(), message.end(), '@', PLAYER_NAME()[0]);
+    ReplaceSpecialCharacters(message);
+    ReplaceColors(message);
+    message += MESSAGE_END();
+}
```
i'm bouncing back and forth on wanting to combine these and have conditional logic within the method for when we have an itemid vs when we don't and wanting to keep them separate (despite a bit of copypasta)

figure starting a thread on it makes sense

leggettc18 replied:
What default value would I set ItemID to? I don't think it can be null, and I'm pretty sure 0 is a valid ItemID (Deku Sticks I think?) -1? This might also have to be refactored again later when the GetItemEntries table gets refactored/enhanced.

briaguya0 replied:
maybe `-1` works? if so then we could have something like
```cpp
void CustomMessageManager::FormatCustomMessage(std::string& message) {
    FormatCustomMessage(message, -1);
}
```

but i'm not sure if it'd end up being cleaner or not, as i said before, bouncing back and forth

leggettc18 replied:
I was thinking something more like:
```c++
void CustomMessageManager::FormatCustomMessage(std::string& message, ItemID iid = -1) {
    if (iid != -1) {
        message.insert(0, ITEM_OBTAINED(iid));
    }
    size_t start_pos = 0;
    std::replace(message.begin(), message.end(), '&', NEWLINE()[0]);
    if (iid != -1) {
        while ((start_pos = message.find('^', start_pos)) != std::string::npos) {
            message.replace(start_pos, 1, WAIT_FOR_INPUT() + ITEM_OBTAINED(iid));
            start_pos += 3;
        }
    } else {
        std::replace(message.begin(), message.end(), '^', WAIT_FOR_INPUT()[0]);
    }
    std::replace(message.begin(), message.end(), '@', PLAYER_NAME()[0]);
    ReplaceSpecialCharacters(message);
    ReplaceColors(message);
    message += MESSAGE_END();
}
```
Not sure if that syntax is exactly right for optionally specifiable arguments with default values in C++, but something like that. `ItemID` also might need to be `uint8_t` or something for that to work, which might be better anyway if we want to use it for other icons.

Alternatively this might even need to go away altogether in favor of some mod-developer friendly way of specifying the icon to show within the strings themselves (which technically they can already do if they know the control codes, that's what the altar text is doing currently).

briaguya0 replied:
actually yeah, replacing the itemid stuff with a more user/modder friendly way to do icons in the strings is definitely the move

leggettc18 replied:
Any suggestions on the syntax? I'm thinking something along the lines of `${iconID}`, but maybe something else would be better?

briaguya0 commented on soh/soh/Enhancements/custom_message/CustomMessageManager.cpp:
```diff
@@ -0,0 +1,150 @@
+}
+
+bool CustomMessageManager::InsertCustomMessage(std::string tableID, uint16_t textID, CustomMessageEntry messages) {
+    auto result = messageTables.find(tableID);
```
not the biggest fan of `result` as the name here, only thing that comes to mind is`tablePair` but i'm not the biggest fan of that either

leggettc18 replied:
Yeah, it's only really used as an intermediary value anyway so I didn't put too much stock in that name.

leggettc18 replied:
I've issued some commits that should resolve most of the renaming suggestions, check them out when you get a chance!

briaguya0 commented on soh/soh/Enhancements/custom_message/CustomMessageManager.cpp:
```diff
@@ -0,0 +1,150 @@
+    FormatCustomMessage(messages.german, iid);
+    FormatCustomMessage(messages.french, iid);
+    const uint16_t textID = giid;
+    return InsertCustomMessage(tableID, textID, messages);
```
the message vs messages stuff is a little confusing here. it seems like in some contexts all of the translations of a message are considered to be one message, where as in some other context they're considered to be multiple messages.

not sure how to best clean it up, just wanted to bring attention to it

leggettc18 replied:
changing that variable to `messageEntry` might make more sense. It is technically only one message, just in multiple languages and that variable also includes the textbox Position and Type.

briaguya0 commented on soh/soh/Enhancements/custom_message/CustomMessageManager.cpp:
```diff
@@ -0,0 +1,150 @@
+}
+
+CustomMessageEntry CustomMessageManager::RetrieveMessage(std::string tableID, uint16_t textID) {
+    std::unordered_map<std::string, CustomMessageTable>::const_iterator result = messageTables.find(tableID);
```
would be nice to have a more descriptive name than result here too

briaguya0 commented on soh/soh/Enhancements/custom_message/CustomMessageManager.cpp:
```diff
@@ -0,0 +1,150 @@
+}
+
+bool CustomMessageManager::ClearMessageTable(std::string tableID) {
+    auto result = messageTables.find(tableID);
```
another result name

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -2407,9 +2403,9 @@ std::string Randomizer::GetGanonHintText() const {
-std::string Randomizer::GetHintFromCheck(RandomizerCheck check) {
-    return this->hintLocations[check];
-}
+//CustomMessageEntry Randomizer::GetHintFromCheck(RandomizerCheck check) {
+//    return CustomMessage::Instance->RetrieveMessage(hintMessageTableID, check);
+//}
```
i assume this is leftover from earlier in dev, just noting it as worth deleting

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -1796,58 +1824,26 @@ std::string AltarIconString(char iconChar) {
         }
     }
-
+    
```
nitpick

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -4745,9 +4741,90 @@ void DrawRandoEditor(bool& open) {
+} GetItemMessage;
+
+void CreateGetItemMessages(std::vector<GetItemMessage> messageEntries) {
+    CustomMessageManager* customMessage = CustomMessageManager::Instance;
```
maybe rename this to customMessageManager or something? having it just named "customMessage" feels like it should be the actual custom message

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -4745,9 +4741,90 @@ void DrawRandoEditor(bool& open) {
+    for (u8 price : prices) {
+        customMessage->CreateMessage(Randomizer::scrubMessageTableID, price,
+            { TEXTBOX_TYPE_BLACK, TEXTBOX_POS_BOTTOM,
+              "\x12\x38\x82\All right! You win! In return for&sparing me, I will sell you a&%gmysterious item%w!&%r" +
```
is there any way we could remove the `\x12\x38\x82\` stuff from these? doesn't need to be a part of this PR but would be nice to have those represented in a more understandable manner

leggettc18 replied:
Yeah at some point we need to branch out some kind of standardized markup language of some kind for replacing all the control codes. That control code is for playing the sound effect, which currently the only custom messages I needed to add to the tables that did that were the Deku Scrubs. So far for that markup-language style thing I borrowed what you already had for the hint text in spoiler files (so `&` for newline, `^` to wait for input, and `@` for the player name. Then I added a %<letter> syntax for changing colors. I proposed a `${<iconID>}` syntax for item icons on the left side of the textbox, maybe we can do a `#{<soundID>}` syntax for sound effects? Or is there something else you'd like to see?

I could also probably just expose some helper methods and people can do string concatenation as an option at least, but I don't want that to be the only way, since people couldn't really use that if we wanted to allow pulling the strings from an external file later on.

briaguya0 replied:
might be worth making an issue for updating our custom text syntax to support icons and sounds, that can be a different PR (especially since i'd like to hear some more opinions on syntax)

also, it raises some other questions (does `&/^` make sense? should we use `%` for colors? can we standardize a bit more without making everything too verbose?)

leggettc18 replied:
We could also potentially use the syntax from Non-JP Dev Files table from here https://wiki.cloudmodding.com/oot/Text_Format#Control_Codes unless that could cause copyright issues. It doesn't have any syntax for line break or changing colors though, so we'd need to come up with something else for those.

briaguya0 replied:
i feel like having our own syntax might be better, especially when thinking about iconIds (they aren't very human readable)

i feel like ideally we'd be able to read a custom text string and understand what it's doing without needing to reference external id tables etc. (trying to balance that with keeping the strings short will be a bit of a challenge, but i'm sure we'll figure something out)

i had an idea for using emoji for text colors (not sure how i feel about using emoji in json but in some ways `🟥` feels better to me than `TEXT_COLOR_RED` or `%r`), so there are definitely some "thinking out of the box" options we can look into

edit: link to comment from colorful hints issue (https://github.com/HarbourMasters/Shipwright/issues/737#issuecomment-1182055293)

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.h:
```diff
@@ -38,9 +43,11 @@ class Randomizer {
     std::string GetGanonText() const;
     std::string GetGanonHintText() const;
-    std::string GetHintFromCheck(RandomizerCheck check);
+    //CustomMessageEntry GetHintFromCheck(RandomizerCheck check);
```
another place to remove instead of comment out

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.h:
```diff
@@ -38,9 +43,11 @@ class Randomizer {
+    //CustomMessageEntry GetHintFromCheck(RandomizerCheck check);
     GetItemID GetRandomizedItemIdFromKnownCheck(RandomizerCheck randomizerCheck, GetItemID ogId);
     GetItemID GetRandomizedItemId(GetItemID ogId, s16 actorId, s16 actorParams, s16 sceneNum);
+    //std::string GetCustomGetItemMessage(GlobalContext* globalCtx, GetItemID giid);
```
delete not comment out

leggettc18 replied:
Done, plus a few other unused functions/definitions that were unused removed as well.

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1479,27 +1495,24 @@ extern "C" int CopyScrubMessage(u16 scrubTextId, char* buffer, const int maxBuff
-                                                   : OTRGlobals::Instance->gRandomizer->GetChildAltarText();
-    return CopyStringToCharBuffer(altarText, buffer, maxBufferSize);
+extern "C" CustomMessageEntry Randomizer_CopyAltarMessage() {
+    return (LINK_IS_ADULT) ? CustomMessageManager::Instance->RetrieveMessage(Randomizer::hintMessageTableID, 0x7088)
+                           : CustomMessageManager::Instance->RetrieveMessage(Randomizer::hintMessageTableID, 0x7040);
```
the hex values here feel a little like magic numbers, any chance we could use an enum or something for the known ones?

leggettc18 replied:
Possibly. I think those particular values match the vanilla `textID`s that they are overriding, but some clarity could be good. Might be a `#define` rather than an enum though since we're calling out select values and with enums the positioning of any later added values would be important.

briaguya0 replied:
with an enum you can do
```cpp
enum blarg {
    a = 3,
    b = 20,
    c = 9001
}
```
so positioning doesn't need to matter

leggettc18 replied:
Well they do have to go in order with enums right? Although that's not exactly difficult to do in code if you're adding an explicit textId anyway. You're right, enum seems cleaner.

leggettc18 replied:
Fixed in 5a97e9f

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1509,3 +1522,90 @@ extern "C" s32 Randomizer_GetRandomizedItemId(GetItemID ogId, s16 actorId, s16 a
+    const int maxBufferSize = sizeof(font->msgBuf);
+    CustomMessageEntry messageEntry;
+    if (gSaveContext.n64ddFlag) {
+        if (textId == 0xF8) {
```
`0xF8` feels like magic, maybe making a macro for it with a descriptive name would be helpful?

leggettc18 replied:
Good idea, 0xF8 is an "unused" `textID` in vanilla (technically it's used but the box only contains the text of the address itself) that fits within the u8 type used for the `textID`s in `sGetItemEntries`.

leggettc18 replied:
Added a file that contains an enum typedef for the raw textIDs that were present in the code in 5a97e9f

briaguya0 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -652,15 +652,15 @@ static GetItemEntry sGetItemTable[] = {
-    GET_ITEM(ITEM_BOTTLE_WITH_BUGS, OBJECT_GI_INSECT, GID_BUG, 0x7A, 0x80, CHEST_ANIM_LONG),
-    GET_ITEM(ITEM_BOTTLE_WITH_POE, OBJECT_GI_GHOST, GID_POE, 0x97, 0x80, CHEST_ANIM_LONG),
-    GET_ITEM(ITEM_BOTTLE_WITH_BIG_POE, OBJECT_GI_GHOST, GID_BIG_POE, 0xF9, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_RED_POTION, OBJECT_GI_LIQUID, GID_POTION_RED, 0xF8, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_GREEN_POTION, OBJECT_GI_LIQUID, GID_POTION_GREEN, 0xF8, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_BLUE_POTION, OBJECT_GI_LIQUID, GID_POTION_BLUE, 0xF8, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_FAIRY, OBJECT_GI_BOTTLE, GID_BOTTLE, 0xF8, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_FISH, OBJECT_GI_FISH, GID_FISH, 0xF8, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_BLUE_FIRE, OBJECT_GI_FIRE, GID_BLUE_FIRE, 0xF8, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_BUGS, OBJECT_GI_INSECT, GID_BUG, 0xF8, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_POE, OBJECT_GI_GHOST, GID_POE, 0xF8, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_BIG_POE, OBJECT_GI_GHOST, GID_BIG_POE, 0xF8, 0x80, CHEST_ANIM_LONG),
```
just another spot where `0xF8` feels like magic, would be nice to either macro it or enum it or something

leggettc18 replied:
Fixed in 5a97e9f

briaguya0 reviewed (changes requested):
Overall this is looking incredible and i'm super excited to see it get in. So much nicer than what we had before.

Left some comments with questions and some comments with just little cleanup notes.

leggettc18 committed:
Moves the NOGDI define to CustomMessageManager.h

leggettc18 committed:
Renames some variables for clarity

leggettc18 committed:
Removes some unneeded function definitions (some of which were already commented out).

leggettc18 committed:
Factors away some  raw textIDs into an enum with descriptive names.

leggettc18 commented:
All right @briaguya-ai I think I've addressed all of your comments except for the one about the modder-friendly syntax for displaying the icons that I wanted your opinion/suggestion on before I started.

leggettc18 committed:
Hopefully fixes build errors on Linux-CI.

leggettc18 commented:
Heh, can't seem to win. Weird linux only error causing builds to fail, after fixing that linux builds fine but jenkins failed to connect to the mac machine lol.

leggettc18 committed:
Cleanup.

leggettc18 commented:
Linux build is failing right now but I'm not sure it has anything to do with any of my code.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -1477,6 +1483,26 @@ void Randomizer::LoadHintLocations(const char* spoilerFileName) {
+    CustomMessageManager::Instance->ClearMessageTable(Randomizer::hintMessageTableID);
+    CustomMessageManager::Instance->AddCustomMessageTable(Randomizer::hintMessageTableID);
+
+    CustomMessageManager::Instance->CreateMessage(
+        Randomizer::hintMessageTableID, 0x7040,
+                                           { TEXTBOX_TYPE_BLUE, TEXTBOX_POS_BOTTOM, gSaveContext.childAltarText,
+                                             gSaveContext.childAltarText, gSaveContext.childAltarText });
+    CustomMessageManager::Instance->CreateMessage(
+        Randomizer::hintMessageTableID, 0x7088,
+                                           { TEXTBOX_TYPE_BLUE, TEXTBOX_POS_BOTTOM, gSaveContext.adultAltarText,
+                                             gSaveContext.adultAltarText, gSaveContext.adultAltarText });
+    CustomMessageManager::Instance->CreateMessage(
+        Randomizer::hintMessageTableID, 0x70CC,
+                                           { TEXTBOX_TYPE_BLACK, TEXTBOX_POS_BOTTOM, gSaveContext.ganonHintText,
+                                             gSaveContext.ganonHintText, gSaveContext.ganonHintText });
+    CustomMessageManager::Instance->CreateMessage(
+        Randomizer::hintMessageTableID, 0x70CD,
+                                           { TEXTBOX_TYPE_BLACK, TEXTBOX_POS_BOTTOM, gSaveContext.ganonText,
+                                             gSaveContext.ganonText, gSaveContext.ganonText });
```
looks like you have the enum values for these in `CustomMessageTypes.h`, let's use them!

leggettc18 replied:
I added them in there because I needed access to them in areas where `#include message_data_static.h` caused type redefinition errors. I suppose if that's all we need from there I can switch the include from `message_data_static.h` to `CustomMessageTypes.h`

leggettc18 replied:
Oh wait I confused myself here. Yes I just missed the raw textIDs here.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -1796,53 +1824,21 @@ std::string AltarIconString(char iconChar) {
-        {"ù", 0x9C },
-        {"û", 0x9D },
-        {"ü", 0x9E }
+        { "À", 0x80 }, { "î", 0x81 }, { "Â", 0x82 }, { "Ä", 0x83 }, { "Ç", 0x84 }, { "È", 0x85 }, { "É", 0x86 },
+        { "Ê", 0x87 }, { "Ë", 0x88 }, { "Ï", 0x89 }, { "Ô", 0x8A }, { "Ö", 0x8B }, { "Ù", 0x8C }, { "Û", 0x8D },
+        { "Ü", 0x8E }, { "ß", 0x8F }, { "à", 0x90 }, { "á", 0x91 }, { "â", 0x92 }, { "ä", 0x93 }, { "ç", 0x94 },
+        { "è", 0x95 }, { "é", 0x96 }, { "ê", 0x97 }, { "ë", 0x98 }, { "ï", 0x99 }, { "ô", 0x9A }, { "ö", 0x9B },
+        { "ù", 0x9C }, { "û", 0x9D }, { "ü", 0x9E }
```
it seems with your changes this stuff exists in custommessagemanager now too. any ideas on how to reduce copypasta here?

leggettc18 replied:
It exists here because this is for pulling the text from the spoilerfile into the saveContext (which then gets pulled into the custom message tables afterwards). I could expose `FormatMessage` instead of it being a private method though.

briaguya0 replied:
i don't think making `FormatMessage` public really hurts anything, but if you run into errors and it ends up taking more than like 10 minutes to switch over to using it then removing this copypasta can be considered out of scope for this pr

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -4745,9 +4737,90 @@ void DrawRandoEditor(bool& open) {
+    std::string english;
+    std::string german;
+    std::string french;
+} GetItemMessage;
```
is there a reason this is defined here or would it make more sense in `CustomMessageTypes.h`? (honest question not sure which makes more sense in this case)

leggettc18 replied:
It might make more sense in `CustomMessageTypes.h`. That file was created after the first batch of reviews from you, due to type redefinition errors when `message_data_static.h` was included in certain files. I pulled the bits from that file I was actually using into `CustomMessageTypes.h`. That particular struct you called out is also likely going to go away if we add support for a syntax for item icons, along with `CreateGetItemMessage` as a whole.

briaguya0 replied:
coolcool, even if we plan on not using it in the future i think it'd be nice to have it in CustomMessageTypes so it doesn't get lost when we refactor away from needing it (`randomizer.cpp` is really big)

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -4745,9 +4737,90 @@ void DrawRandoEditor(bool& open) {
+
+void Randomizer::CreateCustomMessages() {
+    const std::vector<GetItemMessage> getItemMessages = {
+        GIMESSAGE(GI_BOTTLE_WITH_BLUE_FIRE, ITEM_BLUE_FIRE,
+                    "You got a %rBottle with Blue &Fire%w! Use it to melt Red Ice!",
+                    "You got a %rBottle with Blue &Fire%w! Use it to melt Red Ice!",
+                    "You got a %rBottle with Blue &Fire%w! Use it to melt Red Ice!"),
+        GIMESSAGE(GI_BOTTLE_WITH_BIG_POE, ITEM_BIG_POE,
+                    "You got a %rBig Poe in a Bottle%w!&Sell it to the Ghost Shop!",
+                    "You got a %rBig Poe in a Bottle%w!&Sell it to the Ghost Shop!",
+                    "You got a %rBig Poe in a Bottle%w!&Sell it to the Ghost Shop!"),
+        GIMESSAGE(GI_BOTTLE_WITH_BLUE_POTION, ITEM_POTION_BLUE,
+                    "You got a %rBottle of Blue Potion%w!&Drink it to replenish your&%ghealth%w and %bmagic%w!",
+                    "You got a %rBottle of Blue Potion%w!&Drink it to replenish your&%ghealth%w and %bmagic%w!",
+                    "You got a %rBottle of Blue Potion%w!&Drink it to replenish your&%ghealth%w and %bmagic%w!"),
+        GIMESSAGE(GI_BOTTLE_WITH_FISH, ITEM_FISH,
+                    "You got a %rFish in a Bottle%w!&It looks fresh and delicious!&They say Jabu-Jabu loves them!",
+                    "You got a %rFish in a Bottle%w!&It looks fresh and delicious!&They say Jabu-Jabu loves them!",
+                    "You got a %rFish in a Bottle%w!&It looks fresh and delicious!&They say Jabu-Jabu loves them!"),
+        GIMESSAGE(GI_BOTTLE_WITH_BUGS, ITEM_BUG,
+                    "You got a %rBug in a Bottle%w!&They love to burrow in&dirt holes!",
+                    "You got a %rBug in a Bottle%w!&They love to burrow in&dirt holes!",
+                    "You got a %rBug in a Bottle%w!&They love to burrow in&dirt holes!"),
+        GIMESSAGE(GI_BOTTLE_WITH_FAIRY, ITEM_FAIRY, "You got a %rFairy in a Bottle%w!&Use it wisely!",
+                    "You got a %rFairy in a Bottle%w!&Use it wisely!",
+                    "You got a %rFairy in a Bottle%w!&Use it wisely!"),
+        GIMESSAGE(GI_BOTTLE_WITH_RED_POTION, ITEM_POTION_RED,
+                    "You got a %rBottle of Red Potion%w!&Drink it to replenish your&%ghealth%w!",
+                    "You got a %rBottle of Red Potion%w!&Drink it to replenish your&%ghealth%w!",
+                    "You got a %rBottle of Red Potion%w!&Drink it to replenish your&%ghealth%w!"),
+        GIMESSAGE(GI_BOTTLE_WITH_GREEN_POTION, ITEM_POTION_GREEN,
+                    "You got a %rBottle of Green Potion%w!&Drink it to replenish your&%bmagic%w!",
+                    "You got a %rBottle of Green Potion%w!&Drink it to replenish your&%bmagic%w!",
+                    "You got a %rBottle of Green Potion%w!&Drink it to replenish your&%bmagic%w!"),
+        GIMESSAGE(GI_BOTTLE_WITH_POE, ITEM_POE,
+                    "You got a %rPoe in a Bottle%w!&That creepy Ghost Shop might&be interested in this...",
+                    "You got a %rPoe in a Bottle%w!&That creepy Ghost Shop might&be interested in this...",
+                    "You got a %rPoe in a Bottle%w!&That creepy Ghost Shop might&be interested in this..."),
```
doesn't need to happen now, but it might be nice to figure out a way to default to using the english string for all 3 when we don't have translations or something

ideally it'd be easy to go through the codebase and find messages that need to be translated, and i feel like having the english string repeated 3 times may not be the best way to make that happen

leggettc18 replied:
Yeah because of how I did that I could pretty easily add a `#define` that just takes the one message and expands it to a struct that includes that one message 3 times. Call it something that indicates it's only in one language so it can show up in searches, like `GIMESSAGE_UNTRANSLATED` or something.

briaguya0 replied:
> `GIMESSAGE_UNTRANSLATED`

this sounds perfect

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1407,99 +1411,39 @@ extern "C" RandomizerCheck Randomizer_GetCheckFromActor(s16 sceneNum, s16 actorI
-    }
-
-    return CopyStringToCharBuffer(scrubText, buffer, maxBufferSize);
+    return CustomMessageManager::Instance->RetrieveMessage(Randomizer::scrubMessageTableID, price);
```
are we passing in the price as the textId here?

leggettc18 replied:
Yes, I have it right now so the scrubs have a table to themselves and in scrubsanity, scrubs with the same price can just use the same messageEntry. Definitely subject to change though.

briaguya0 replied:
it does feel a little weird to me but i makes sense that it works, might be worth making a code cleanup issue to look into ways to make it clearer

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1407,99 +1411,39 @@ extern "C" RandomizerCheck Randomizer_GetCheckFromActor(s16 sceneNum, s16 actorI
-    const std::string& altarText = (LINK_IS_ADULT) ? OTRGlobals::Instance->gRandomizer->GetAdultAltarText()
-                                                   : OTRGlobals::Instance->gRandomizer->GetChildAltarText();
-    return CopyStringToCharBuffer(altarText, buffer, maxBufferSize);
+extern "C" CustomMessageEntry Randomizer_CopyAltarMessage() {
```
might be worth renaming this since it's not copying the altar message anymore, but instead retrieving/returning the custom message entry

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1407,99 +1411,39 @@ extern "C" RandomizerCheck Randomizer_GetCheckFromActor(s16 sceneNum, s16 actorI
-extern "C" int Randomizer_CopyGanonText(char* buffer, const int maxBufferSize) {
-    const std::string& ganonText = OTRGlobals::Instance->gRandomizer->GetGanonText();
-    return CopyStringToCharBuffer(ganonText, buffer, maxBufferSize);
+extern "C" CustomMessageEntry Randomizer_CopyGanonText() {
```
same copy naming thing here

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1407,99 +1411,39 @@ extern "C" RandomizerCheck Randomizer_GetCheckFromActor(s16 sceneNum, s16 actorI
-extern "C" int Randomizer_CopyGanonHintText(char* buffer, const int maxBufferSize) {
-    const std::string& ganonText = OTRGlobals::Instance->gRandomizer->GetGanonHintText();
-    return CopyStringToCharBuffer(ganonText, buffer, maxBufferSize);
+extern "C" CustomMessageEntry Randomizer_CopyGanonHintText() {
```
copy naming

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1407,99 +1411,39 @@ extern "C" RandomizerCheck Randomizer_GetCheckFromActor(s16 sceneNum, s16 actorI
 
-extern "C" int Randomizer_CopyHintFromCheck(RandomizerCheck check, char* buffer, const int maxBufferSize) {
-    // we don't want to make a copy of the std::string returned from GetHintFromCheck
+extern "C" CustomMessageEntry Randomizer_CopyHintFromCheck(RandomizerCheck check) {
```
copy naming

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1509,3 +1453,89 @@ extern "C" s32 Randomizer_GetRandomizedItemId(GetItemID ogId, s16 actorId, s16 a
+    const int maxBufferSize = sizeof(font->msgBuf);
+    CustomMessageEntry messageEntry;
+    if (gSaveContext.n64ddFlag) {
+        if (textId == TEXT_RANDOMIZER_GI_ESCAPE_HATCH) {
```
`TEXT_RANDOMIZER_GI_ESCAPE_HATCH` is a really fun name, but it'd be nice to try to come up with something a little more descriptive. i'm not sure what a better name would be but figured it's worth noting that it was unclear to me when reading what it is supposed to be/do

leggettc18 replied:
I thought it was pretty descriptive, as it describes the text id used to "escape" from the usual code for getting messages during get-item scenario into the custom text system. Would `TEXT_RANDOMIZER_CUSTOM_ITEM` be a better choice?

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1509,3 +1453,89 @@ extern "C" s32 Randomizer_GetRandomizedItemId(GetItemID ogId, s16 actorId, s16 a
+        if (textId == TEXT_RANDOMIZER_GI_ESCAPE_HATCH) {
+            messageEntry =
+                Randomizer_GetCustomGetItemMessage((GetItemID)GET_PLAYER(globalCtx)->getItemId, buffer, maxBufferSize);
+        } else if (textId == 0x2053 && Randomizer_GetSettingValue(RSK_GOSSIP_STONE_HINTS) != 0 &&
```
`0x2053` is a magic number here, it'd be nice to get it into the enum

leggettc18 replied:
Yep, think there's already an enum with that ID, but if not I can add one. Must've just missed that one.

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1509,3 +1453,89 @@ extern "C" s32 Randomizer_GetRandomizedItemId(GetItemID ogId, s16 actorId, s16 a
+            } else {
+                textId = TEXT_GS_FREEZE;
+            }
+            messageEntry = CustomMessageManager::Instance->RetrieveMessage("BaseGameOverrides", textId);
```
i'm not sure how i feel about strings for tableIds, but i can see using strings for tableId as a way to enable easier custom tables with modding in the future.

might be good to put the "always there" type strings into an enum or defines or something in `CustomMessageTypes.h` to avoid needing to update string literals in multiple places if we change our minds about how to name something

leggettc18 replied:
I do already have something in place for rando at least, with static variables on the Randomizer class holding the IDs for its message tables. I think I might have just forgotten to replace this one after verifying it worked.

briaguya0 commented on soh/soh/z_message_OTR.cpp:
```diff
@@ -1,18 +1,22 @@
 extern "C" MessageTableEntry* sStaffMessageEntryTablePtr;
 //extern "C" MessageTableEntry* _message_0xFFFC_nes;	
 
+const std::string customMessageTableID = "BaseGameOverrides";
```
this goes along with the last comment i made, maybe just moving this to `CustomMessageTypes.h` would be good?

leggettc18 replied:
The idea is that when mods come around they will keep track of their own message table IDs in a similar way, so hardcoding that in the custom message system itself doesn't quite feel right.

briaguya0 replied:
yeah, not sure where the best place for it would be. i think i've just dealt with a lot of string literals needing to match issues lately and i'm hoping to find a way to turn things into compile time errors instead of runtime errors

leggettc18 replied:
Yeah the idea is that when these are initially defined is the only time you'll ever see a string literal. That one from one of your other comments was just something I forgot to replace with the variable I think. I might also be able to refactor it to dynamically give a UUID whenever you create a message table and store that in the class instance that generated it somehow, but that's still requiring strings to match at the end of the day.

briaguya0 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -652,15 +653,15 @@ static GetItemEntry sGetItemTable[] = {
-    GET_ITEM(ITEM_BOTTLE_WITH_BUGS, OBJECT_GI_INSECT, GID_BUG, 0x7A, 0x80, CHEST_ANIM_LONG),
-    GET_ITEM(ITEM_BOTTLE_WITH_POE, OBJECT_GI_GHOST, GID_POE, 0x97, 0x80, CHEST_ANIM_LONG),
-    GET_ITEM(ITEM_BOTTLE_WITH_BIG_POE, OBJECT_GI_GHOST, GID_BIG_POE, 0xF9, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_RED_POTION, OBJECT_GI_LIQUID, GID_POTION_RED, TEXT_RANDOMIZER_GI_ESCAPE_HATCH, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_GREEN_POTION, OBJECT_GI_LIQUID, GID_POTION_GREEN, TEXT_RANDOMIZER_GI_ESCAPE_HATCH, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_BLUE_POTION, OBJECT_GI_LIQUID, GID_POTION_BLUE, TEXT_RANDOMIZER_GI_ESCAPE_HATCH, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_FAIRY, OBJECT_GI_BOTTLE, GID_BOTTLE, TEXT_RANDOMIZER_GI_ESCAPE_HATCH, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_FISH, OBJECT_GI_FISH, GID_FISH, TEXT_RANDOMIZER_GI_ESCAPE_HATCH, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_BLUE_FIRE, OBJECT_GI_FIRE, GID_BLUE_FIRE, TEXT_RANDOMIZER_GI_ESCAPE_HATCH, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_BUGS, OBJECT_GI_INSECT, GID_BUG, TEXT_RANDOMIZER_GI_ESCAPE_HATCH, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_POE, OBJECT_GI_GHOST, GID_POE, TEXT_RANDOMIZER_GI_ESCAPE_HATCH, 0x80, CHEST_ANIM_LONG),
+    GET_ITEM(ITEM_BOTTLE_WITH_BIG_POE, OBJECT_GI_GHOST, GID_BIG_POE, TEXT_RANDOMIZER_GI_ESCAPE_HATCH, 0x80, CHEST_ANIM_LONG),
```
it's not clear what `TEXT_RANDOMIZER_GI_ESCAPE_HATCH` does in this context either, just a note that whatever name we land on should help clarify what's happening when someone reads these lines too

briaguya0 reviewed (changes requested):
lots of small comments, some that might make sense as new issues outside this pr, some that are quite small and would be good to address before merging, and some that are just starting conversations

overall this is looking awesome and i'm super excited for it. i think it's really close to ready!

leggettc18 committed:
Adds newlines to the ends of some files.

leggettc18 committed:
Addresses most of bria's comments.

leggettc18 committed:
Exposes FormatCustomMessage and uses it in FormatJsonHintText.

leggettc18 commented:
@briaguya-ai Everything but the syntax for replacing symbols with control codes that we wanted more opinions on should have been addressed now in the most recent batch of commits.

leggettc18 committed:
Replaces string literal for BaseGameOverrides with const variable.

briaguya0 reviewed (approved):
beautiful!

# :shipit:

briaguya0 commented:
created an issue for custom message syntax https://github.com/HarbourMasters/Shipwright/issues/1002

after thinking about the ability to have options choices (like "do you want me to repeat that?" yes/no stuff), i started thinking it might make sense to use json objects instead of just doing things as strings

that would definitely open up some possibilities

leggettc18 commented:
JSON objects? Can you illustrate what the JSON would look like? I'm having trouble understanding what you mean.

leggettc18 committed:
Merge changes from develop-zhora

leggettc18 committed:
Adds documentation comments to CustomMessageManager.h

leggettc18 committed:
Merge branch 'develop-zhora' of https://github.com/HarbourMasters/Shipwright into custom-messages

Kenix3 commented:
Merge pending conflict resolution.

leggettc18 commented:
Appreciate it! I'm having some issues with unresolved external symbols while trying to merge in the cmake changes as we speak. Not sure why but I'll got to Discord for help with that.

leggettc18 committed:
Adapts this branch for cmake building.

leggettc18 commented:
Alright, got cmake working. If Jenkins builds OK we should be good.
